src/zarr/codecs/batched_codec_pipeline.py

```diff
@@ -0,0 +1,439 @@
+from __future__ import annotations
+
+from itertools import islice
+from typing import TYPE_CHECKING, Iterator, TypeVar, Iterable
+from warnings import warn
+import numpy as np
+from dataclasses import dataclass
+
+from zarr.config import config
+from zarr.abc.codec import (
+    ByteGetter,
+    ByteSetter,
+    Codec,
+    CodecPipeline,
+    ArrayArrayCodec,
+    ArrayBytesCodec,
+    ArrayBytesCodecPartialDecodeMixin,
+    ArrayBytesCodecPartialEncodeMixin,
+    BytesBytesCodec,
+)
+from zarr.codecs.registry import get_codec_class
+from zarr.common import JSON, concurrent_map, parse_named_configuration
+from zarr.indexing import is_total_slice
+from zarr.metadata import ArrayMetadata
+
+if TYPE_CHECKING:
+    from typing_extensions import Self
+    from zarr.common import ArraySpec, BytesLike, SliceSelection
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+
+def _unzip2(iterable: Iterable[tuple[T, U]]) -> tuple[list[T], list[U]]:
+    out0: list[T] = []
+    out1: list[U] = []
+    for item0, item1 in iterable:
+        out0.append(item0)
+        out1.append(item1)
+    return (out0, out1)
+
+
+def batched(iterable: Iterable[T], n: int) -> Iterable[tuple[T, ...]]:
+    if n < 1:
+        raise ValueError("n must be at least one")
+    it = iter(iterable)
+    while batch := tuple(islice(it, n)):
+        yield batch
+
+
+def resolve_batched(codec: Codec, chunk_specs: Iterable[ArraySpec]) -> Iterable[ArraySpec]:
+    return [codec.resolve_metadata(chunk_spec) for chunk_spec in chunk_specs]
+
+
+@dataclass(frozen=True)
+class BatchedCodecPipeline(CodecPipeline):
+    """Default codec pipeline.
+
+    This batched codec pipeline divides the chunk batches into batches of a configurable
+    batch size ("mini-batch"). Fetching, decoding, encoding and storing are performed in
+    lock step for each mini-batch. Multiple mini-batches are processing concurrently.
+    """
```

Comment:
this can come later but we're going to want some additional documentation on the behavior here. Reading this, I'm not entirely sure I get it.